### PR TITLE
Julia v0.5

### PR DESCRIPTION
**Pull request by bicycle1885**
This is a pull request to make a step forward to the next Julia v0.5 and ultimately drop the support of the current Julia v0.4.

At the moment, to pass all tests, you will need to
* <del>build the latest master branch of Julia</del> Julia 0.5 RC2, and
* <del>clone the latest master branch of <del>Libz.jl</del> (use v0.2) and LightXML.jl</del> install the lastest Libz.jl and LightXML.jl

### CODE REVIEW

**Comment by bicycle1885** [comment]
I'd like to merge this change as soon as possible because I want to merge https://github.com/bicycle1885/HTSFileFormats.jl into Bio.jl. But with this change, Bio.jl no longer works on Julia 0.4 so you developers need to switch your Julia compiler to Julia 0.5.

**Comment by bicycle1885** [comment]
@Ward9250 @jgreener64 @kdmurray91 will be affected by this change since you have open pull request.

**Comment by Ward9250** [comment]
I'm happy to switch to 0.5, code in the PRs I'm working on _shouldnt_ (famous last words 😆 ) be affected but if they are I'll simply address the issues that crop up. I heard tell that 0.5 is release now? Or Travis nightly now refers to 0.6??

**Comment by bicycle1885** [comment]
Pull requests that add new features *shouldn't* be affected but you may need to reconsider your code because many functions and types are deprecated in 0.5. Julia 0.5 is not yet released, but according to their words, it will reach an official release within a few weeks.

**Comment by bicycle1885** [comment]
Julia 0.5 seems to have significant performance regression on `BioSequence`.

```julia
using Bio.Seq

function count_a(seq)
    n = 0
    for nt in seq
        n += nt == DNA_A
    end
    return n
end

count_a(randdnaseq(10))
seq = randdnaseq(10_000_000)
@time count_a(seq)
@time count_a(seq)
@time count_a(seq)

println("--- baseline ---")
count_a(collect(randdnaseq(10)))
seq = collect(seq)
@time count_a(seq)
@time count_a(seq)
@time count_a(seq)
```

In this simple benchmark, Julia 0.5 was 7x slower than Julia 0.4.
```
~/.j/v/Bio (julia-v0.5|…) $ julia test.jl
  0.018301 seconds (149 allocations: 10.167 KB)
  0.019434 seconds (5 allocations: 176 bytes)
  0.019703 seconds (5 allocations: 176 bytes)
--- baseline ---
  0.010130 seconds (5 allocations: 176 bytes)
  0.008483 seconds (5 allocations: 176 bytes)
  0.009051 seconds (5 allocations: 176 bytes)
~/.j/v/Bio (julia-v0.5|…) $ julia5 test.jl
WARNING: Method definition require(Symbol) in module Base at loading.jl:317 overwritten in
 module Main at /Users/kenta/.julia/v0.5/Requires/src/require.jl:12.
  0.134248 seconds (134 allocations: 7.781 KB)
  0.137695 seconds (5 allocations: 176 bytes)
  0.135939 seconds (5 allocations: 176 bytes)
--- baseline ---
  0.008715 seconds (5 allocations: 176 bytes)
  0.008731 seconds (5 allocations: 176 bytes)
  0.009581 seconds (5 allocations: 176 bytes)
```

File: src/seq/kmer.jl [hunk 1]
```diff
@@ -131,14 +131,12 @@ end
...
-@compat begin
-    Base.:-{T,K}(x::Kmer{T,K}, y::Integer) =
-        Kmer{T,K}(UInt64(x) - reinterpret(UInt64, Int64(y)))
-    Base.:+{T,K}(x::Kmer{T,K}, y::Integer) =
-        Kmer{T,K}(UInt64(x) + reinterpret(UInt64, Int64(y)))
-    Base.:+{T,K}(x::Integer, y::Kmer{T,K}) = y + x
-    Base.:(==){T,k}(x::Kmer{T,k}, y::Kmer{T,k}) = UInt64(x) == UInt64(y)
-end
+Base.:-{T,K}(x::Kmer{T,K}, y::Integer) =
+    Kmer{T,K}(UInt64(x) - reinterpret(UInt64, Int64(y)))
```
Comment (vchuravy) [inline src/seq/kmer.jl]: This can be written as `UInt64(x) - y%UInt64`

File: src/seq/quality.jl [hunk 2]
```diff
@@ -92,7 +92,7 @@ function infer_quality_encoding(data::Vector{UInt8}, start, stop,
                                 default::QualityEncoding=EMPTY_QUAL_ENCODING)
     @inbounds for i in start:stop
         c = data[i]
-        if '!' <= c <= '~'
+        if UInt8('!') <= c <= UInt8('~')
```
Comment (vchuravy) [inline src/seq/quality.jl]: similarily `'!' % UInt8` is slightly faster.

File: src/seq/kmer.jl (hunk 1 above)
Comment (bicycle1885) [inline src/seq/kmer.jl]: Looks nice! Fixed.

**Comment by kdmurray91** [comment]
@bicycle1885 If at all possible (and only if it won't delay your work or inconvenience you) I'd like to hold off on merging this till the final 0.5 release binaries are provided. I like the idea of having master compatible with whatever one downloads from  the "Current Release" header on the [download page](http://julialang.org/downloads/). Perhaps this is a case for a "development" or "future" branch that is 0.5-only?

File: src/seq/quality.jl (hunk 2 above)
Comment (bicycle1885) [inline src/seq/quality.jl]: I'd like to deprecate this quality encoding inference feature in the next release. So I'd like to keep this as is for now.

**Comment by bicycle1885** [comment]
@kdmurray91 I agree with you, but I'm now working on #265, which depends on this pull request, to finish my project, so I'd like to merge this change as soon as possible. Switching between 0.4 and 0.5 on my local development branch is a little bit frustrating and I want to focus my effort on a single Julia version. Fortunately, the release of 0.5 would be very soon (I guess, within a week or two) since the next release candidate version is the third one.

**Comment by kdmurray91** [comment]
OK, do what you feel is best!

**Comment by bicycle1885** [comment]
Yes, 0.5 points to the latest RC (I read it somewhere, but I don't recall). 0.5 will pass all the tests but there is at least one performance regression mentioned above.

**Comment by bicycle1885** [comment]
Julia 0.5 RC3 has been released: <https://groups.google.com/forum/#!topic/julia-users/AhH-porwS4o>. So, I think it's time to merge this pull request if the running CI says okay.

**Comment by Ward9250** [comment]
Sounds good to me!

**Comment by kdmurray91** [comment]
:+1:

**Pull request merged**